community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DiskLayer.java

```diff
@@ -179,70 +183,164 @@ public int labelGetForName( String label )
+                        }
+                        else
+                        {
+                            throw new NoSuchElementException();
+                        }
+                    }
+                };
+            }
+            else
             {
-                if ( !hasNext() )
-                {
-                    throw new NoSuchElementException();
-                }
-                return safeCastLongToInt( labels[cursor++] );
+                throw new EntityNotFoundException( EntityType.NODE, nodeId );
             }
-        };
+        }
     }
 
     @Override
-    public RelationshipIterator nodeListRelationships( long nodeId, Direction direction )
+    public RelationshipIterator nodeListRelationships( StoreStatement statement,
+            long nodeId,
+            Direction direction )
             throws EntityNotFoundException
     {
-        return StoreRelationshipIterable.iterator( neoStore, nodeId, alwaysTrue(), direction );
+        return nodeListRelationships( statement, nodeId, direction, null );
     }
 
     @Override
-    public RelationshipIterator nodeListRelationships( long nodeId, Direction direction, int[] relTypes )
+    public RelationshipIterator nodeListRelationships( final StoreStatement statement,
+            long nodeId,
+            Direction direction,
+            int[] relTypes )
             throws EntityNotFoundException
     {
-        // TODO Instead of having a PrimitiveIntSet here we can (in the dense case) have a sorted relTypes array
-        // and use a predicate that knows that the types coming from accept will be sorted as well.
-        return StoreRelationshipIterable.iterator( neoStore, nodeId, asSet( relTypes ), direction );
+        try ( final StoreSingleNodeCursor nodeCursor = statement.acquireSingleNodeCursor() )
+        {
+            if ( nodeCursor.init( nodeId ).next() )
+            {
+                final RelationshipCursor relCursor = nodeCursor.get().relationships( direction, relTypes );
+
+                return new RelationshipIterator()
+                {
+                    boolean hasNext = relCursor.next();
+
+                    long id;
+                    int type;
+                    long startNode;
+                    long endNode;
+
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return hasNext;
+                    }
+
+                    @Override
+                    public long next()
+                    {
+                        if ( hasNext )
+                        {
+                            try
+                            {
+                                RelationshipValue value = relCursor.get();
+                                id = value.getId();
+                                type = value.getType();
+                                startNode = value.getStartNode();
+                                endNode = value.getEndNode();
```

Comment:
If we need to do this, I see no value in having the `RelationshipValue` interface